### PR TITLE
AEA-1001, AEA-1003 remove all distributed packages and add them to local registry, make …

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR moves all distributed packages into the local registry. It makes the necessary changes to accommodate this change. Two hacks are introduced: for default decision maker and error handling the packages are lazy loaded.

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/aea.py [hunk 1]
```diff
@@ -229,49 +236,59 @@ def get_multiplexer_setup_options(self) -> Optional[Dict]:
...
+
+    def _handle_decoding(
+        self, envelope: Envelope, protocol: Optional[Protocol]
+    ) -> Tuple[Optional[Message], List[Handler]]:
         error_handler = self._get_error_handler()
 
-        if error_handler is None:
-            self.logger.warning("ErrorHandler not initialized. Stopping AEA!")
-            raise _StopRuntime()
+        from packages.fetchai.skills.error.handlers import (  # noqa: F811 # pylint: disable=import-outside-toplevel
```
Comment (DavidMinarsch) [inline aea/aea.py:261]: hack

File: aea/cli/interact.py [hunk 2]
```diff
@@ -34,17 +34,24 @@
...
 from aea.identity.base import Identity
 from aea.mail.base import Envelope, Message
 from aea.multiplexer import InBox, Multiplexer, OutBox
-from aea.protocols.default.dialogues import DefaultDialogue, DefaultDialogues
-from aea.protocols.default.message import DefaultMessage
 from aea.protocols.dialogue.base import Dialogue as BaseDialogue
+from aea.protocols.dialogue.base import Dialogues
+
+
+if TYPE_CHECKING:
```
Comment (DavidMinarsch) [inline aea/cli/interact.py:44]: hack

File: aea/decision_maker/default.py [hunk 3]
```diff
@@ -38,14 +38,25 @@
...
-from aea.protocols.signing.dialogues import SigningDialogues as BaseSigningDialogues
-from aea.protocols.signing.message import SigningMessage
-from aea.protocols.state_update.dialogues import StateUpdateDialogue
-from aea.protocols.state_update.dialogues import (
-    StateUpdateDialogues as BaseStateUpdateDialogues,
-)
-from aea.protocols.state_update.message import StateUpdateMessage
+
+
+if TYPE_CHECKING:
```
Comment (DavidMinarsch) [inline aea/decision_maker/default.py:43]: hack

File: aea/helpers/file_io.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Read to and write from file with envelopes."""
+
```
Comment (DavidMinarsch) [inline aea/helpers/file_io.py:22]: extracted from stub connection for general use

File: docs/agent-vs-aea.md [hunk 5]
```diff
@@ -19,7 +19,7 @@ from aea.connections.base import Connection
 from aea.connections.stub.connection import StubConnection, write_with_lock
 from aea.identity.base import Identity
 from aea.mail.base import Envelope
-from aea.protocols.default.message import DefaultMessage
+from packages.fetchai.protocols.default.message import DefaultMessage
```
Comment (DavidMinarsch) [inline docs/agent-vs-aea.md:22]: still need to fix docs

File: packages/fetchai/connections/stub/connection.py [hunk 6]
```diff
@@ -24,117 +24,28 @@
...
 INPUT_FILE_KEY = "input_file"
 OUTPUT_FILE_KEY = "output_file"
-DEFAULT_INPUT_FILE_NAME = "./input_file"
-DEFAULT_OUTPUT_FILE_NAME = "./output_file"
 SEPARATOR = b","
 
 PUBLIC_ID = PublicId.from_str("fetchai/stub:0.11.0")
 
 
-def _encode(e: Envelope, separator: bytes = SEPARATOR):
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/stub/connection.py:49]: taken out and moved to helpers

File: scripts/generate_api_docs.py [hunk 7]
```diff
@@ -86,10 +86,10 @@
     "aea.protocols.generator.common": "api/protocols/generator/common.md",
     "aea.protocols.generator.extract_specification": "api/protocols/generator/extract_specification.md",
     "aea.protocols.generator.validate": "api/protocols/generator/validate.md",
-    "aea.protocols.default.custom_types": "api/protocols/default/custom_types.md",
-    "aea.protocols.default.dialogues": "api/protocols/default/dialogues.md",
-    "aea.protocols.default.message": "api/protocols/default/message.md",
-    "aea.protocols.default.serialization": "api/protocols/default/serialization.md",
+    "packages.fetchai.protocols.default.custom_types": "api/protocols/default/custom_types.md",
```
Comment (DavidMinarsch) [inline scripts/generate_api_docs.py:89]: fix

File: aea/aea.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/aea.py:261]: Re handler: the tricky bit is using the handle(message)  interface.

File: aea/aea.py [hunk 8]
```diff
@@ -51,13 +56,15 @@
 from aea.helpers.logging import AgentLoggerAdapter, get_logger
```
Comment (marcofavorito) [inline aea/aea.py:56]: move the imports `aea.decision_maker.*` below, where `DecisionMakerHandler` and `DefaultDecisionMakerHandler` are actually used

File: tests/test_cli/test_fetch.py [hunk 9]
```diff
@@ -150,7 +150,7 @@ class TestFetchFromRemoteRegistry(AEATestCaseMany):
     @pytest.mark.flaky(reruns=MAX_FLAKY_RERUNS)
     def test_fetch_agent_from_remote_registry_positive(self):
         """Test fetch agent from Registry for positive result."""
-        self.run_cli_command("fetch", "fetchai/my_first_aea:0.12.0")
+        self.run_cli_command("fetch", "fetchai/my_first_aea:0.13.0")
```
Comment (DavidMinarsch) [inline tests/test_cli/test_fetch.py:153]: probably revert

File: aea/cli/interact.py [hunk 10]
```diff
@@ -63,7 +71,39 @@ def interact(click_context: click.core.Context):  # pylint: disable=unused-argum
...
+            "protocols",
+            STATE_UPDATE_PROTOCOL.name,
+        )
+    )
+    Connection.from_dir(
+        Path(
+            "vendor", DEFAULT_CONNECTION.author, "connections", DEFAULT_CONNECTION.name
+        ),
+        agent_identity,
+        None,
```
Comment (marcofavorito) [inline aea/cli/interact.py:95]: we don't need the crypto store for StubConnection

Comment (marcofavorito) [reply aea/cli/interact.py]: updated

**Review by marcofavorito — APPROVED** [review]
LGTM

File: aea/cli/add.py [hunk 11]
```diff
@@ -116,7 +116,7 @@ def add_item(ctx: Context, item_type: str, item_public_id: PublicId) -> None:
     ctx.clean_paths.append(dest_path)
 
     is_distributed = is_distributed_item(item_public_id)
-    if is_local and is_distributed:
+    if is_local and is_distributed:  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/add.py:119]: Could remove code, kept for now

File: aea/aea_builder.py [hunk 12]
```diff
@@ -954,7 +969,7 @@ def _get_decision_maker_handler_class(self) -> Type[DecisionMakerHandler]:
         return (
             self._decision_maker_handler_class
             if self._decision_maker_handler_class is not None
-            else self.DEFAULT_DECISION_MAKER_HANDLER_CLASS
+            else None
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:972]: simplify

File: aea/cli/interact.py [hunk 13]
```diff
@@ -149,29 +219,24 @@ def _construct_message(action_name: str, envelope: Envelope):
...
-            "Provide message of protocol fetchai/default:0.7.0 for performative {}:".format(
-                performative_str
-            )
+            f"Provide message of protocol '{str(message_class.protocol_id)}' for performative {performative_str}:"
         )
         message_escaped = input()  # nosec
         message_escaped = message_escaped.strip()
         if message_escaped == "":
             raise InterruptInputException
-        if performative == DefaultMessage.Performative.BYTES:
```
Comment (DavidMinarsch) [inline aea/cli/interact.py:168]: hardcoded, can never get to else

File: aea/cli/publish.py [hunk 14]
```diff
@@ -108,7 +108,7 @@ def _save_agent_locally(ctx: Context) -> None:
         dependencies = getattr(ctx.agent_config, item_type_plural)
         for public_id in dependencies:
             if public_id in DISTRIBUTED_PACKAGES:
-                continue
+                continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:111]: could remove, kept for now

File: aea/cli/utils/decorators.py [hunk 15]
```diff
@@ -155,7 +155,7 @@ def _cast_ctx(context: Union[Context, click.core.Context]) -> Context:
     """
     if isinstance(context, Context):
         return context
-    if isinstance(context, click.core.Context):
+    if isinstance(context, click.core.Context):  # pragma: no cover
```
Comment (DavidMinarsch) [inline aea/cli/utils/decorators.py:158]: currently not used, kept for reuse in future

File: aea/cli/utils/package_utils.py [hunk 16]
```diff
@@ -329,7 +329,7 @@ def find_item_in_distribution(  # pylint: disable=unused-argument
             "Cannot find {} with author and version specified.".format(item_type)
         )
 
-    return package_path
+    return package_path  # pragma: no cover
```
Comment (DavidMinarsch) [inline aea/cli/utils/package_utils.py:332]: kept for now, currently not reached

File: aea/cli/interact.py [hunk 17]
```diff
@@ -33,18 +34,34 @@
...
-from aea.configurations.loader import ConfigLoader
-from aea.connections.stub.connection import (
-    DEFAULT_INPUT_FILE_NAME,
-    DEFAULT_OUTPUT_FILE_NAME,
-    StubConnection,
+from aea.configurations.constants import (
+    DEFAULT_CONNECTION,
+    DEFAULT_PROTOCOL,
+    SIGNING_PROTOCOL,
+    STATE_UPDATE_PROTOCOL,
```
Comment (solarw) [inline aea/cli/interact.py:41]: would be nice to have one constant to have a set of all default packages,  so maintain easily, package add/load  mechanism can be updated to support the variable amount of components with respect to package type

**Review by solarw — APPROVED** [review]
LGTM

**Pull request merged**